### PR TITLE
Reorganize and clarify API doc generator

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This cleans up a lot of issues with the API doc generation.

Here are the main changes:
 - Rename `dartdoc.dart` to `create_api_docs.dart`
 - Move the bulk of the operations out of `dev/bots/docs.sh` into `create_api_docs.dart`.
 - Delete `dashing_postprocess.dart` and `java_and_objc.dart` and incorporate those operations into `create_api_docs.dart`.
 - Refactor the doc generation into more understandable classes
 - Bump the snippets tool version to 0.4.0 (the latest one)
 - Centralize the information gathering about the Flutter repo into the new `FlutterInformation` class.
 - Clean up the directory handling, and convert to using the `file` package for all file and directory paths.
 - Add an `--output` option to docs.sh that specifies the location of the output ZIP file containing the docs.
   - Defaults to placing the output in `dev/docs/api_docs.zip` (i.e. where the previous code generates the file).
 - Moved all document generation into a temporary folder that is removed once the documents are generated, to avoid VSCode and other IDEs trying to index the thousands of HTML and JS files in the docs output.
 - Updated pubspec dependencies.

## Tests
 - Added tests for doc generation.

### CODE REVIEW

File: dev/tools/pubspec.yaml [hunk 1]
```diff
@@ -45,7 +46,6 @@ dev_dependencies:
   node_preamble: 2.0.2 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   package_config: 2.1.0 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
   pool: 1.5.1 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
-  pub_semver: 2.1.4 # THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade"
```
Comment (Hixie) [inline dev/tools/pubspec.yaml:48]: to make this change did you use `flutter update-packages`?

**Review by Hixie — APPROVED** [review]
LGTM

**Review by godofredoc — APPROVED** [review]
Anything to consider or pay attention to on release candidate branches and stable?

File: dev/tools/pubspec.yaml (hunk 1 above)
Comment (gspencergoog) [reply dev/tools/pubspec.yaml:48]: Oh, good point.  I'll re-run that.

**Comment by gspencergoog** [comment]
> Anything to consider or pay attention to on release candidate branches and stable?

This should generate the same `dev/docs/api_docs.zip` file as before, but I haven't tested the actual site deployment part, where it unpacks that ZIP file and uploads the site. How would I do that?

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by christopherfujino** [comment]
@gspencergoog you'll probably ahve to rebase upstream and re-run `flutter update-packages --force-upgrade` after https://github.com/flutter/flutter/commit/f0e7c518164167638d55ec848365f1aa8a7cb6b2 was merged.

**Pull request merged**